pkg/handlers/utils/secrets.go

```diff
@@ -64,30 +66,62 @@ func CopySecretToRemoteCluster(
 	return nil
 }
 
-// EnsureOwnerReferenceForSecret will ensure that the secretName Secret has an OwnerReference of the cluster.
-func EnsureOwnerReferenceForSecret(
+// EnsureClusterOwnerReferenceForObject ensures that OwnerReference of the cluster is added on provided object.
+func EnsureClusterOwnerReferenceForObject(
 	ctx context.Context,
 	cl ctrlclient.Client,
-	secretName string,
+	objectRef *corev1.TypedLocalObjectReference,
 	cluster *clusterv1.Cluster,
 ) error {
-	secret, err := getSecretForCluster(ctx, cl, secretName, cluster)
+	targetObj, err := GetResourceFromTypedLocalObjectReference(
+		ctx,
+		cl,
+		objectRef,
+		cluster.Namespace,
+	)
 	if err != nil {
 		return err
 	}
 
-	err = controllerutil.SetOwnerReference(cluster, secret, cl.Scheme())
+	err = controllerutil.SetOwnerReference(cluster, targetObj, cl.Scheme())
 	if err != nil {
-		return fmt.Errorf("failed to set owner reference on Secret: %w", err)
+		return fmt.Errorf("failed to set cluster's owner reference on object: %w", err)
 	}
 
-	err = cl.Update(ctx, secret)
+	err = cl.Update(ctx, targetObj)
 	if err != nil {
-		return fmt.Errorf("failed to update Secret with owner references: %w", err)
+		return fmt.Errorf("failed to update object with cluster's owner reference: %w", err)
 	}
 	return nil
 }
 
+// GetResourceFromTypedLocalObjectReference gets the resource from the provided TypedLocalObjectReference.
+func GetResourceFromTypedLocalObjectReference(
+	ctx context.Context,
+	cl ctrlclient.Client,
+	typedLocalObjectRef *corev1.TypedLocalObjectReference,
+	ns string,
+) (*unstructured.Unstructured, error) {
+	apiVersion := corev1.SchemeGroupVersion.String()
+	if typedLocalObjectRef.APIGroup != nil {
+		apiVersion = *typedLocalObjectRef.APIGroup
+	}
```

Comment:
Since it is generic function not sure if should default apiVersion to `v1`.
incase we want to default to v1 then following would read better.
```suggestion
	apiVersion := *typedLocalObjectRef.APIGroup
	if apiVersion  == nil {
		apiVersion = corev1.SchemeGroupVersion.String()
	}
```

Reply:
I don't think it would panic, Shalin's suggestion is just flipping the if condition to make it a little more obvious its falling back to the default.

Reply:
+1. I am just asking to make it obvious.

Reply:
Ahh I see what you mean by nil pointer https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/actions/runs/13165433472/job/36744307943, because of that deref it fails.

Reply:
Wait - this will still panic if `typedLocalObjectRef` is `nil`.

Reply:
```suggestion
	apiVersion := corev1.SchemeGroupVersion.String()
	if typedLocalObjectRef != nil {
		apiVersion = *typedLocalObjectRef.APIGroup
	}
```

Reply:
Good point @jimmidyson, lets be safer here no need to pass a pointer https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/1034/commits/e78131c5313b48f3d925cfaf69a4229d3eeea6ef